### PR TITLE
Add initial implementation of `PackageWatcher`.

### PR DESCRIPTION
**Pull request by matanlurey**
Purposefully did not try to change any internals of `build_runner` to use it (yet).

Work towards #456.

/cc @natebosch

### CODE REVIEW

File: build_runner/lib/src/watcher/package_watcher.dart [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// Actual (default) implementation that delegates to DirectoryWatcher.
+class _PackageWatcher implements PackageWatcher {
+  static Watcher _defaultWatch(String path) => new DirectoryWatcher(path);
+
+  final _CreateWatcher _createWatcher;
+  final PackageGraph _packageGraph;
+
+  const _PackageWatcher(this._packageGraph,
+      {_CreateWatcher watch: _defaultWatch})
+      : _createWatcher = watch;
```
Comment (jakemac53) [inline build_runner/lib/src/watcher/package_watcher.dart]: nit: don't use the default args cause they suck, just check for null in the assignment:

```dart
 const _PackageWatcher(this._packageGraph,
      {_CreateWatcher watch})
      : _createWatcher = watch ?? _defaultWatch;
```

File: build_runner/lib/src/watcher/package_watcher.dart [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+typedef Watcher _CreateWatcher(String path);
+
+// Actual (default) implementation that delegates to DirectoryWatcher.
+class _PackageWatcher implements PackageWatcher {
+  static Watcher _defaultWatch(String path) => new DirectoryWatcher(path);
+
+  final _CreateWatcher _createWatcher;
+  final PackageGraph _packageGraph;
+
+  const _PackageWatcher(this._packageGraph,
```
Comment (jakemac53) [inline build_runner/lib/src/watcher/package_watcher.dart]: Shouldn't this really take a `Package` not a `PackageGraph`? We need to be able to watch packages other than the root one.

Comment (matanlurey) [reply build_runner/lib/src/watcher/package_watcher.dart]: In the next PR, I'm making the following change:

```dart
@protected
Stream watchPath(String path);

Stream watchGraph();
```

... which sort of begs this to be called `PackageGraphWatcher`. Does that make sense?

Comment (jakemac53) [reply build_runner/lib/src/watcher/package_watcher.dart]: ah ok, so `watchPath` only watches the root, and `watchGraph` watches all packages?

Comment (matanlurey) [reply build_runner/lib/src/watcher/package_watcher.dart]: Err, `watchNode(PackageNode node)`, is probably better than `watchPath`.

The idea is `watchGraph()` calls `watchNode(...)` for each tracked dependency/node.

Comment (jakemac53) [reply build_runner/lib/src/watcher/package_watcher.dart]: Ah ok cool, sgtm :)

**Comment by matanlurey** [comment]
Closing in favor of a better approach.

**Pull request closed without merging**